core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -84,10 +84,14 @@
   private boolean hasNewDeleteFiles = false;
 
   MergingSnapshotProducer(String tableName, TableOperations ops) {
+    this(tableName, ops, ops.current().spec());
+  }
+
+  MergingSnapshotProducer(String tableName, TableOperations ops, PartitionSpec spec) {
```

Comment:
I don't think that `spec` should be passed in the constructor. See my comments on `BaseRewriteFiles`.

Instead, I think we should initially set the spec ID to null and initialize it the first time a `DataFile` is added. Then validate that all other data files use the same ID. That way we don't need a new API method and maintain flexibility.